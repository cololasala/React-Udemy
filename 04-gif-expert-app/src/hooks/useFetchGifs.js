import { useEffect, useState } from 'react';
import { getGifs } from '../helpers/getGifs';

export const useFetchGifs = (category) => { // este custom hook sirve  para esperar las images y controlar el loading

    const [state, setState] = useState({
        data: [],
        loading: true
    });

    useEffect(() => {           // controlo para que getGifs sea ejecutado solo 1 vez, el array vacio es un array de dependencias
        getGifs(category)
            .then(imgs =>
                setTimeout(() => {
                    setState({
                        data: imgs,
                        loading: false
                    });
                })              // se le saca 3000 seg para que ande el useFetchGifs.test.js
            )
            .catch(error => console.log(error));
    }, [category]);         // si la category cambia se ejecutara otra vez el useEffect

    return state;

}