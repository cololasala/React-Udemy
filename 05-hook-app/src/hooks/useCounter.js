import React, { useState } from 'react'

export const useCounter = (initialState = 10) => {  // si no paso nada se inicializa por defecto con 10
    const [state, setState] = useState(initialState);

    const increment = () => {
        setState(state + 1);
    }

    const decrement = (factor = 1) => {
        setState(state - factor);
    }

    const reset = () => {
        setState(initialState);
    }

    return {                    // retorno la logica de mi custom hook en este componente
        state,
        increment,
        decrement,
        reset
    }
}
