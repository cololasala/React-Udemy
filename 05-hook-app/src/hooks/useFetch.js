import { useState } from "react";
import { useEffect } from "react";

export const useFetch = (url) => {

    const [state, setState] = useState({
        data: null,                             // data que viene del fetch
        isLoading: true,                       // para saber si esta cargando
        hasErrors: null
    })

    const getFetch = async () => {
        setState({
            ...state,
            isLoading: true
        });

        const res = await fetch(url);
        const data = await res.json();

        setState({
            data: data,
            isLoading: false,
            hasErrors: null
        });
        console.log(data)
    }

    useEffect(() => {
        getFetch();

        return () => {}
    }, [url]);           // cada vez que el url cambie se volvera a disparar el useEffect

    return {
        data: state.data,                             
        isLoading: state.isLoading,                       
        hasErrors: state.hasErrors
    }
}
