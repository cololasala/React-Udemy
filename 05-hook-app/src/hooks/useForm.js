import React, { useState } from 'react'

export const useForm = (initialForm = {}) => {                  // inicio de form objeto vacio
    const [stateForm, setStateForm] = useState(initialForm);

    const handleInputChange = ({ target }) => { // desestructuro el event
        const { name, value } = target;
        setStateForm({
            ...stateForm,
            [name]: value             // target name seria "name" o "email" (depende que input lo llame) y target value seria el valor propio del input
        })
    }

    const onResetForm = () => {
        setStateForm(
            initialForm
        );
    }

    return {
        stateForm,
        handleInputChange,
        onResetForm
    }
}
