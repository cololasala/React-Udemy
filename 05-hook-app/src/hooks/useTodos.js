import { useReducer, useEffect } from "react";
import { todoReducer } from '../components/08-useReducer/todoReducer';

const initialState = [];
const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || [];
}

export const useTodos = () => {
    const [todos, dispatch] = useReducer(todoReducer, initialState, init); // el primer argumento es nuestro Reducer, el segundo es nuestro initialState, el tercero seria nuestra funcion inicializadora (para casos pesados, en nuestro caso para inicializar con lo del localStorage)

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]);

    const onNewTodo = (newTodo) => {
        const action = {
            type: '[TODO] Add todo',
            payload: newTodo,
        }

        dispatch(action);       // con el dispatch paso la action al reducer
    }

    const onDeleteTodo = (id) => {
        const action = {
            type: '[TODO] Remove todo',
            payload: id,
        }

        dispatch(action);
    }

    const onToggleTodo = (id) => {
        const action = {
            type: '[TODO] Toggle todo',
            payload: id
        }

        dispatch(action)
    }

    const pendingsTodos = () => {
        return todos.filter((todo) => todo.done === false).length;
    }

    return {
        todos,
        onNewTodo,
        onDeleteTodo,
        onToggleTodo,
        pendingsTodos,
        todosCount: todos.length
    }
}
