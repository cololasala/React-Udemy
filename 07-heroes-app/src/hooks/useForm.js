import { useState } from 'react';

export const useForm = (initialForm = {}) => {

    const [formState, setFormState] = useState(initialForm);

    const onInputChange = ({ target }) => {
        const { name, value } = target;
        setFormState({
            ...formState,
            [name]: value
        });
    }

    const onResetForm = () => {
        setFormState(initialForm);
    }

    return {
        ...formState, // retorno todas las properties del objeto form (en este caso seria textSearch)
        formState,
        onInputChange,
        onResetForm,
    }
}