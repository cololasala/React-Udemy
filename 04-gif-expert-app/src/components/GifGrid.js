import GifItem from './GifItem';
import { useFetchGifs } from '../hooks/useFetchGifs';
import PropTypes from 'prop-types';

const GifGrid = ({ category }) => {

   // const [images, setImages] = useState([]);
    const { data: images, loading } = useFetchGifs(category);

    //useEffect(() => {       // controlo para que getGifs sea ejecutado solo 1 vez, el array vacio es un array de dependencias
    //    getGifs(category)
    //        .then( imgs => setImages(imgs)) // otra manera es solo ponerlo asi -> setImages (solo eso funciona igual)
    //        .catch(error => console.log(error));
    //}, [category]);         // si la category cambia se ejecutara otra vez el useEffect

    return (
        <>
            <h3>{category}</h3>
            { loading && <p className='animate__animated animate__flash'>Loading...</p> } {/* sin tengo loading entonces muestro la etiqueta */}

            <div className='card-grid'>
                {
                    images.map(img =>
                        <GifItem
                            key={img.id}
                            {...img} />
                    )
                }
            </div>
        </>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}

export default GifGrid;