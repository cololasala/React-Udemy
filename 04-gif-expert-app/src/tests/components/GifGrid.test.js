import { shallow } from 'enzyme';
import GifGrid from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');              // permite fingir cualquier llamada a useFetchGifs y controlar la info que responde

describe('prueba <GifGrid />', () => {

    test('debe crearse correctamente', () => {
        useFetchGifs.mockReturnValue({          // permite el retorno de los valores de mi useFetchGifs
            data: [],
            loading: true
        });

        const category = '';
        const wrapper = shallow(<GifGrid category={category} />);

        expect(wrapper).toMatchSnapshot();
    });

    test('debe mostrar items cuando se cargan imagenes usando useFetchGifs', () => {
        const gifs = [{
            id: 'abc',
            url: 'http://localhost:2000/cualquiercosa.jpg',
            title: 'nada'
        }];

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });

        const category = '';
        const wrapper = shallow(<GifGrid category={category} />);

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('p').exists()).toBe(false);
        expect(wrapper.find('GifItem').length).toBe(gifs.length);
    });
})