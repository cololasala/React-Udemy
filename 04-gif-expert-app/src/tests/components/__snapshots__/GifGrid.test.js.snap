// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prueba <GifGrid /> debe crearse correctamente 1`] = `
<Fragment>
  <h3 />
  <p
    className="animate__animated animate__flash"
  >
    Loading...
  </p>
   
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`prueba <GifGrid /> debe mostrar items cuando se cargan imagenes usando useFetchGifs 1`] = `
<Fragment>
  <h3 />
   
  <div
    className="card-grid"
  >
    <GifItem
      id="abc"
      key="abc"
      title="nada"
      url="http://localhost:2000/cualquiercosa.jpg"
    />
  </div>
</Fragment>
`;
